### PR TITLE
Elasticsearch stores elasticsearch for indexes with history deletion …

### PR DESCRIPTION
**Pull request by yeerpeng**
Elasticsearch stores indexes can be configured to create indexes on a day-by-day basis as specified

Please answer these questions before submitting pull request

- Why submit this pull request?
Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.
Elasticsearch stores elasticsearch for indexes with history deletion capability: create indexes on a daily basis as specified by configuration and delete indexes directly on a daily basis

### CODE REVIEW

**Comment by peng-yongsheng** [comment]
@yeerpeng 

1. This class ElasticSearchUtil.java has nothing to do with Elasticsearch, why is it called ElasticSearchUtil.java.
2. Dynamically build index names in multiple methods(ElasticSearchUtil, createIndex, formatIndexName, deleteHistory......).
3. There is no other storage implementation.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java [hunk 1]
```diff
@@ -186,7 +187,7 @@ public CoreModuleProvider() {
 
         DataTTLKeeperTimer.INSTANCE.setDataTTL(moduleConfig.getDataTTL());
         DataTTLKeeperTimer.INSTANCE.start(getManager());
-
+        ElasticSearchUtil.INSTANCE.setRecordDataTTL(moduleConfig.getRecordDataTTL());
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/CoreModuleProvider.java]: Why core module changed?

File: oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/base/StorageEsInstaller.java [hunk 2]
```diff
@@ -79,6 +80,9 @@ public StorageEsInstaller(ModuleManager moduleManager, int indexShardsNumber, in
 
     @Override protected void createTable(Client client, Model tableDefine) throws StorageException {
         ElasticSearchClient esClient = (ElasticSearchClient)client;
+        if ((!tableDefine.isDeleteHistory() || tableDefine.getName().endsWith(Downsampling.Month.getName()) || tableDefine.getName().endsWith(Downsampling.Day.getName())) && esClient.getCreateByDayIndexes().contains(tableDefine.getName())) {
```
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/base/StorageEsInstaller.java]: This condition checks look like complex, especially when you based on table name, do you have better way to do that?

File: oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/query/TraceQueryEsDAO.java [hunk 3]
```diff
@@ -125,7 +127,9 @@ public TraceBrief queryBasicTraces(long startSecondTB, long endSecondTB, long mi
         sourceBuilder.query(QueryBuilders.termQuery(SegmentRecord.TRACE_ID, traceId));
         sourceBuilder.size(20);
 
-        SearchResponse response = getClient().search(SegmentRecord.INDEX_NAME, sourceBuilder);
+        ElasticSearchClient client = getClient();
+        String[] indexes = client.getIndexNameByDate(SegmentRecord.INDEX_NAME, 0, 0);
```
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/query/TraceQueryEsDAO.java]: `0, 0` is not a friendly way to say all indexes. You should redesign the API to have the index name list.

**Comment by wu-sheng** [comment]
> There is no other storage implementation.

@peng-yongsheng I think we could leave that to other storage implementations themselves.

**Comment by yeerpeng** [comment]
@wu-sheng @peng-yongsheng 
Thank you very much for your detailed explanation, but now I have some confusion, structure of system configuration TTL configuration in the core module, for example, is now in the store at the same time also want to use this configuration module, even though the code now have these in store module configuration, but seems to be invalid, don't know the recent project team will achieve this piece, I need to change the place on system structure there is a big change

In the future, I may encounter the same problem in other places. :)

**Comment by wu-sheng** [comment]
Each module provider has its own config. TTL is nothing about you want to build the indexes daily. You could choose only to make TTL works in daily if the user says so.

We are more talking about codes consistency and understandable, such as how to get the right index name list.

**Comment by yeerpeng** [comment]
> Each module provider has its own config. TTL is nothing about you want to build the indexes daily. You could choose only to make TTL works in daily if the user says so.
> 
> We are more talking about codes consistency and understandable, such as how to get the right index name list.

It does not matter when it is created, but it does matter when it is queried, and you can only get the exact index names of the indexes created on a daily basis through TTL

**Comment by yeerpeng** [comment]
For example, how many days are left after deleting this interval segment can only be determined by TTL configuration item

**Comment by yeerpeng** [comment]
@wu-sheng @peng-yongsheng 
Hello, two friends. I have adjusted the unreasonable code before, and I hope you can spare some time to give me some advice. Thank you very much    ：）

**Comment by wu-sheng** [comment]
Hi, Peng

I hope you have time to review, I flight to SF today. Can't do review in next nearly two days.

**Comment by wu-sheng** [comment]
@yeerpeng Sorry to keep this open for a long time. @peng-yongsheng is working on a global solution, as we found the metric delete may trigger a crash too, when resource is limited. He will try to submit a pull request in 6.1 or 6.2

I have to close this. Sorry.

**Comment by yeerpeng** [comment]
@wu-sheng Don't mention it. I am looking forward to your perfect solution and wish skywalking more and more development ：）

**Pull request closed without merging**